### PR TITLE
moved ml/core implementation to source

### PR DESCRIPTION
**Pull request by AlexTeslenko**
moved ml/core implementation to source

### CODE REVIEW

File: libs/ml/include/ml/core/node.hpp [hunk 1]
```diff
@@ -153,258 +153,5 @@ class Node
   std::shared_ptr<ops::Ops<TensorType>> op_ptr_;
```
Comment (baykaner) [inline libs/ml/include/ml/core/node.hpp:157]: There seems to be several methods you have not moved over to source file

File: libs/ml/include/ml/core/graph.hpp [hunk 2]
```diff
@@ -275,747 +275,6 @@ std::string Graph<TensorType>::AddNode(std::string const &             node_name
   return updated_name;
```
Comment (baykaner) [inline libs/ml/include/ml/core/graph.hpp:270]: you should move as many #includes from hpp to cpp as possible. Use forward declarations where necessary

**Review by baykaner — CHANGES_REQUESTED** [review]

File: libs/ml/include/ml/core/graph.hpp [hunk 3]
```diff
@@ -275,747 +275,6 @@ std::string Graph<TensorType>::AddNode(std::string const &             node_name
...
 }
 
-/**
```
Comment (iryna-melnyk) [inline libs/ml/include/ml/core/graph.hpp:279]: comment to line 244: I guess AddNode should be also moved to cpp file, right?

Comment (baykaner) [reply libs/ml/include/ml/core/graph.hpp:279]: some of them are extremely difficult or impossible to move (e.g. functions that include a parameter pack will be in general not possible to move, and templated functions will be extremely verbose to move)

File: libs/ml/include/ml/core/node.hpp (hunk 1 above)
Comment (AlexTeslenko) [reply libs/ml/include/ml/core/node.hpp:157]: Maybe it is better to leave such one-line methods in header file and just make them inline?

File: libs/ml/include/ml/core/graph.hpp [hunk 4]
```diff
@@ -18,20 +18,9 @@
 //------------------------------------------------------------------------------
 
 #include "ml/core/node.hpp"
-#include "ml/meta/ml_type_traits.hpp"
 #include "ml/ops/weights.hpp"
 #include "ml/state_dict.hpp"
```
Comment (baykaner) [inline libs/ml/include/ml/core/graph.hpp:23]: I think you can forward declare this and just include it in the source file?

File: libs/ml/include/ml/core/graph.hpp [hunk 5]
```diff
@@ -18,20 +18,9 @@
 //------------------------------------------------------------------------------
 
 #include "ml/core/node.hpp"
-#include "ml/meta/ml_type_traits.hpp"
 #include "ml/ops/weights.hpp"
```
Comment (baykaner) [inline libs/ml/include/ml/core/graph.hpp]: I think weights are not used anywhere in the header - so can be deleted or moved to source

File: libs/ml/include/ml/core/node.hpp [hunk 6]
```diff
@@ -20,16 +20,6 @@
 #include "logging/logging.hpp"
 #include "ml/ops/ops.hpp"
 #include "ml/ops/weights.hpp"
```
Comment (baykaner) [inline libs/ml/include/ml/core/node.hpp]: i think weights include can be moved to source? I don't see any usage of weights in the header

File: libs/ml/src/core/graph.cpp [hunk 7]
```diff
@@ -19,7 +19,18 @@
 #include "math/tensor/tensor.hpp"
 #include "math/tensor/tensor_slice_iterator.hpp"
 #include "ml/core/graph.hpp"
-
+#include "ml/layers/normalisation/layer_norm.hpp"
```
Comment (baykaner) [inline libs/ml/src/core/graph.cpp]: this looks like it definitely should not be here

File: libs/ml/src/core/graph.cpp [hunk 8]
```diff
@@ -19,7 +19,18 @@
...
+#include "ml/meta/ml_type_traits.hpp"
+#include "ml/ops/embeddings.hpp"
```
Comment (baykaner) [inline libs/ml/src/core/graph.cpp]: also this should not be here

**Review by baykaner — CHANGES_REQUESTED** [review]

File: libs/ml/include/ml/core/graph.hpp (hunk 4 above)
Comment (AlexTeslenko) [reply libs/ml/include/ml/core/graph.hpp:23]: If I just forward declare StateDict, then I get this error:
error: implicit instantiation of undefined template 'fetch::ml::StateDict<fetch::math::Tensor<signed char, fetch::memory::SharedArray<signed char, 1> > >'
  virtual fetch::ml::StateDict<TensorType> StateDict();

Cannot figure out if it is possible to fix it without including "ml/state_dict.hpp"
                                           ^

File: libs/ml/include/ml/core/graph.hpp (hunk 5 above)
Comment (AlexTeslenko) [reply libs/ml/include/ml/core/graph.hpp]: If I remove "ml/ops/weights" then I need include these headers:

 #include "ml/regularisers/regulariser.hpp"
#include "ml/exceptions/exceptions.hpp"

**Review by baykaner — APPROVED** [review]

**Pull request merged**